include/zephyr/net/dhcpv4.h

```diff
@@ -44,6 +44,121 @@ enum net_dhcpv4_state {
+ *
+ * These enumerations represent RFC2131 defined msy type codes, hence
+ * they should not be renumbered.
+ *
+ * Additions, removald and reorders in this definition must be reflected
+ * within corresponding changes to net_dhcpv4_msg_type_name.
+ */
+enum net_dhcpv4_msg_type {
+	NET_DHCPV4_MSG_TYPE_DISCOVER	= 1,
+	NET_DHCPV4_MSG_TYPE_OFFER	= 2,
+	NET_DHCPV4_MSG_TYPE_REQUEST	= 3,
+	NET_DHCPV4_MSG_TYPE_DECLINE	= 4,
+	NET_DHCPV4_MSG_TYPE_ACK		= 5,
+	NET_DHCPV4_MSG_TYPE_NAK		= 6,
+	NET_DHCPV4_MSG_TYPE_RELEASE	= 7,
+	NET_DHCPV4_MSG_TYPE_INFORM	= 8,
+};
+
+#ifdef CONFIG_NET_DHCPV4_OPTION_CALLBACKS
+
+struct net_dhcpv4_option_callback;
+
+/**
+ * @typedef net_dhcpv4_option_callback_handler_t
+ * @brief Define the application callback handler function signature
+ *
+ * @param cb Original struct net_dhcpv4_option_callback owning this handler
+ * @param length The length of data returned by the server. If this is
+ *		 greater than cb->max_length, only cb->max_length bytes
+ *		 will be available in cb->data
+ * @param msg_type Type of DHCP message that triggered the callback
+ * @param iface The interface on which the DHCP message was received
+ *
+ * Note: cb pointer can be used to retrieve private data through
+ * CONTAINER_OF() if original struct net_dhcpv4_option_callback is stored in
+ * another private structure.
+ */
+typedef void (*net_dhcpv4_option_callback_handler_t)(struct net_dhcpv4_option_callback *cb,
+						     size_t length,
+						     enum net_dhcpv4_msg_type msg_type,
+						     struct net_if *iface);
+
+/**
+ * @brief DHCP option callback structure
+ *
+ * Used to register a callback in the DHCPv4 client callback list.
+ * As many callbacks as needed can be added as long as each of them
+ * are unique pointers of struct net_dhcpv4_option_callback.
+ * Beware such structure should not be allocated on stack.
+ *
+ * Note: To help setting it, see net_dhcpv4_init_option_callback() below
+ */
+struct net_dhcpv4_option_callback {
+	/** This is meant to be used internally and the user should not
+	 * mess with it.
+	 */
+	sys_snode_t node;
+
+	/** Actual callback function being called when relevant. */
+	net_dhcpv4_option_callback_handler_t handler;
+
+	/** The DHCP option this callback is attached to. */
+	uint8_t option;
+
+	/** Maximum length of data buffer. */
+	size_t max_length;
+
+	/** Pointer to a buffer of size max_length that is used to store the
+	 * option data.
+	 */
+	void *data;
+};
+
+/**
+ * @brief Helper to initialize a struct net_dhcpv4_option_callback properly
+ * @param callback A valid Application's callback structure pointer.
+ * @param handler A valid handler function pointer.
+ * @param option The DHCP option the callback responds to.
+ * @param data A pointer to a buffer for max_length bytes.
+ * @param max_length The maximum length of the data returned.
+ */
+static inline void net_dhcpv4_init_option_callback(struct net_dhcpv4_option_callback *callback,
+						   net_dhcpv4_option_callback_handler_t handler,
```

Comment:
If the callback struct is "public" then the handler parameter is kind of unnecessary as it is part of callback anyway. But I would suggest the keep this prototype like this but make `struct net_dhcpv4_option_callback` private instead.

Reply:
Since this function is used to populate only the fields that should be changed by the user application, the whole point is to call it with a callback pointer and all the data that will be placed inside.
This is how things are also handled for the GPIO callbacks: https://github.com/zephyrproject-rtos/zephyr/blob/d01780fc9403166d250ded3d02372a7e1a9db4fd/include/zephyr/drivers/gpio.h#L1440

Reply:
Doesn't work, since the user code then doesn't know the storage size of this struct. And I can't think of any non-hacky way around that.

Reply:
Fair enough, I was just wondering whether we need to fully expose the struct internals. At least we should mark the callback struct internal as `/** @cond INTERNAL_HIDDEN */ .... /** @endcond */` so that the documentation generator does not try to document the internal fields.

Reply:
I'm fine with that and will update it.